src/darker/verification.py

```diff
@@ -67,3 +67,39 @@ def verify_ast_unchanged(
     except AssertionError as exc_info:
         debug_dump(black_chunks, edited_linenums)
         raise NotEquivalentError(str(exc_info))
+
+
+class ASTVerifier:
+    """
+    Verifies if reformatted TextDocument is AST-equivalent to baseline.
+
+    Keeps in-memory data about previous comparisons to improve performance.
+    """
+
+    def __init__(self, baseline: TextDocument) -> None:
+        self._baseline_ast_str = self._to_ast_str(baseline)
+        self._comparisons: Dict[str, bool] = {baseline.string: True}
+
+    @staticmethod
+    def _to_ast_str(document: TextDocument) -> str:
+        return "\n".join(stringify_ast(parse_ast(document.string)))
+
+    def is_equivalent_to_baseline(self, document: TextDocument) -> bool:
+        """Returns true if document is AST-equivalent to baseline"""
+        if document.string in self._comparisons:
+            return self._comparisons[document.string]
+
+        try:
+            document_ast_str = self._to_ast_str(document)
+        except SyntaxError:
+            comparison = False
+        else:
+            comparison = self._baseline_ast_str == document_ast_str
+
+        self._comparisons[document.string] = comparison
+        return self._comparisons[document.string]
```

Comment:
There are two syntactically identical `return` statements. Would you consider the following a simplification and an improvement or not?
```suggestion
        if document.string not in self._comparisons:
            try:
                document_ast_str = self._to_ast_str(document)
            except SyntaxError:
                comparison = False
            else:
                comparison = self._baseline_ast_str == document_ast_str
            self._comparisons[document.string] = comparison
        return self._comparisons[document.string]
```

Reply:
This is purely opinion-based, but I prefer flatter structure and early returns.

Or maybe we should decouple caching and computing and just use `functools.lru_cache(max_size=None)`?

Reply:
I agree that double returns are a fair trade-off for flatness and an early return.

Separating caching would indeed be more clear, although the current version isn't bad either – caching can't get much simpler than this.

Is `@lru_cache` flushed when it's used on a method and the object goes out of scope? I mean if Darker is processing a huge repository of Python files with lots of changes, we wouldn't like it to keep this cache for files already processed. I didn't find an answer in [lru_cache documentation](https://docs.python.org/3/library/functools.html#functools.lru_cache).

Reply:
>Is @lru_cache flushed when it's used on a method and the object goes out of scope?

Hmm, apparently not:
```python
from functools import lru_cache

class CachedAdder:
    def __init__(self, a):
        self._a = a

    @lru_cache(maxsize=None)
    def add(self, b):
        return self._a + b
    
def test_adder():
    adder = CachedAdder(21)
    print(adder.add(21))
    
print(CachedAdder.add.cache_info())
test_adder()
print(CachedAdder.add.cache_info())
```
```
CacheInfo(hits=0, misses=0, maxsize=None, currsize=0)
42
CacheInfo(hits=0, misses=1, maxsize=None, currsize=1)
```